# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: gfx/2d/InlineTranslator.h
# Commit: 71b6060b6015
# Full Hash: 71b6060b60151e157e94110d28fb51d76bb6415a
# Author: Bob Owen <bobowencode@gmail.com>
# Date: 2023-12-08 04:18:28
# Regressor Bug: 1863914
# File Overlap Count: 1
# Description:
#   Bug 1863914: Use multiple shmem buffers for remote canvas recording. r=aosmond
#   
#   This replaces the use of a single large ring buffer.
#   The buffers are still processed in parallel and are recycled to reduce
#   allocation. Events that do not fit in the default sized buffer have a separate
# ==============================================================================

diff -r bf28b3c066e6 -r 71b6060b6015 gfx/2d/InlineTranslator.h
--- a/gfx/2d/InlineTranslator.h	Thu Dec 07 17:10:06 2023 +0000
+++ b/gfx/2d/InlineTranslator.h	Thu Dec 07 17:27:24 2023 +0000
@@ -166,6 +166,28 @@
   std::string GetError() { return mError; }
 
  protected:
+  // an istream like class for reading from memory
+  struct MemReader {
+    constexpr MemReader(char* aData, size_t aLen)
+        : mData(aData), mEnd(aData + aLen) {}
+    void read(char* s, std::streamsize n) {
+      if (n <= (mEnd - mData)) {
+        memcpy(s, mData, n);
+        mData += n;
+      } else {
+        // We've requested more data than is available
+        // set the Reader into an eof state
+        SetIsBad();
+      }
+    }
+    bool eof() { return mData > mEnd; }
+    bool good() { return !eof(); }
+    void SetIsBad() { mData = mEnd + 1; }
+
+    char* mData;
+    char* mEnd;
+  };
+
   RefPtr<DrawTarget> mBaseDT;
   Matrix mBaseDTTransform;
   nsRefPtrHashtable<nsPtrHashKey<void>, DrawTarget> mDrawTargets;